3155: Create EVM Trace Specification (#3155)
ethereum/EIPs: EIPS/eip-3155.md

---
eip: 3155
title: EVM trace specification
author: Martin Holst Swende (@holiman), Marius van der Wijden (@MariusVanDerWijden)
discussions-to: https://ethereum-magicians.org/t/eip-3155-create-evm-trace-specification/5007
status: Draft
type: Standards Track

> category (*only required for Standard Track): Interface

Reviewer: ```suggestion
category: Interface
```

created: 2020-12-07
---

> <!--You can leave these HTML comments in your merged EIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new EIPs. Note that an EIP number will be assigned by an editor. When opening a pull request to submit your EIP, please use an abbreviated title in the filename, `eip-draft_title_abbrev.md`. The title should be 44 characters or less.-->

Reviewer: ```suggestion
```

## Simple Summary

> <!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->

Reviewer: ```suggestion
```

> This EIP introduces a new JSON standard for EVM traces. There has been a de-facto standard which is used by go-ethereum, OpenEthereum, Besu and Nethermind. This document shall document this standard as well as provide stricter guidelines to make new implementations easier.

Reviewer: I think everything except the first sentence can be moved to motivation. Also, generally don't need to refer to "this EIP" in the simple summary as it is implied.

Author: moved

> This EIP introduces a new JSON standard for EVM traces. There has been a de-facto standard which is used by go-ethereum, OpenEthereum, Besu and Nethermind. This document shall document this standard as well as provide stricter guidelines to make new implementations easier.

Reviewer 2: Not a de-facto standard, other tracing formats are supported by OE, Besu and Nethermind that are friendlier to smart contract developers.  It is a common format.

## Motivation

> <!--The motivation is critical for EIPs that want to change the Ethereum protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the EIP solves. EIP submissions without sufficient motivation may be rejected outright.-->

Reviewer: ```suggestion
```

> The Ethereum Virtual Machine executes all smart contract code on ethereum. In order to easier debug smart contracts and state tests, a de-facto standard was introduced to log every execution step of the EVM. This standard was implemented by go-ethereum, parity and nethermind and Besu. Since the de-facto standard was not well defined, the implementations differed slightly making it hard to develop adequate tooling.

Reviewer: ```suggestion
The Ethereum Virtual Machine executes all smart contract code on ethereum. In order to more easily debug smart contracts and state tests, a de-facto standard was introduced to log every execution step of the EVM. This standard was implemented by go-ethereum, parity and nethermind and Besu. Since the de-facto standard was not well defined, the implementations differed slightly making it hard to develop adequate tooling.
```

> The Ethereum Virtual Machine executes all smart contract code on ethereum. In order to easier debug smart contracts and state tests, a de-facto standard was introduced to log every execution step of the EVM. This standard was implemented by go-ethereum, parity and nethermind and Besu. Since the de-facto standard was not well defined, the implementations differed slightly making it hard to develop adequate tooling.

Reviewer 2: replace de-facto standard with common format.

This EIP has multiple objectives:
- Move the specification to a more visible place to encourage new clients to implement it
- Strictly define corner cases that were not addressed in the previous version
- Allow for updates to the specification in case new fields are introduced during execution
- Provide sample output

Implementing this EIP in all major clients allows us to create meaningful differential fuzzers that fuzz EVM implementations for the mainnet and all upcoming hardforks.
It also helps finding differences in execution quickly in the case of a chain split.

Due to the slight differences in output the following fields currently need to be dropped when comparing tracing outputs: $GasCost$, $MemorySize$, $RefundCounter$, $ReturnStack$, and $ReturnData$. This reduces the usefulness of tracing significantly. Creating a common standard would allow to add these fields thus increasing the surface that can be tested by a fuzzer.

## Specification

> <!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->

Reviewer: ```suggestion
```

Clients should be able to execute simple transactions as well as code and return traces. In the following, we will call this client CUT (client under test) and use go-ethereums `evm` binary for code examples. 

### Inputs

> The CUT MUST be able to execute a statetest as described [here](https://www.ethdocs.org/en/latest/contracts-and-transactions/ethereum-tests/state_tests/index.html).

Reviewer 2: We are blurring together the output of the trace with the way it is invoked.  I would prefer taking all references to the CUT out.  This format is also part of the JSON-RPC calls `debug_standardTraceBlockToFile` and `debug_standardTraceBadBlockToFile` which are not CLI clients and don't have standard error or standard out.  That would make this EIP just specifying the trace line format.  A separate EIP for using this for fuzzing with detail about the CUT could then be opened.

Example :
``` 
evm --json --nomemory statetest foobar.json 
```
The CUT MUST spit out jsonl as defined below, and it MUST also report the final `stateRoot` hash after executing the test. The CUT SHOULD report whether the tests succeeded or failed.

If the statetest contains a Fork identifier (e.g. `Berlin` or `Istanbul`) the CUT MUST run the statetest under the chain rules of this fork.

The CUT SHOULD executed statetests if the filenames are written to STDIN. This allows to execute multiple tests on the same instance without having to set up and tear down the client.

> ### Output

Reviewer 2: Rather than output this would be the trace objects.  Some expository about how this would be an array of objects if used in a wrapping json object like a JSON-RPC call should be added.

> The CUT MUST output a `json` object for EACH operation. The output SHOULD to be printed to STDERR.

Reviewer 2: I don't think STDERR is appropriate for non-error output.

> Required Fields: 

Reviewer 2: What is expected for each field?  a Json number?  a hex encoded string? a plain string? Base64 encoded data?  Those need to be specified as those are the source of some of the incompatibility.

My votes based on current implementation:
Json number - `pc`, `op`, `depth`, `memSize`
Hex encoded number - `gas`, `gasCost`, `refund`, as an array `stack`
Hex encoded String - `returnData`, `error` (error includes revert reason if supported by clients), `memory`
plain string - `opName`
Object keyed and valued by hex object - `storage`

- `pc` Program Counter
- `op` OpCode
- `gas` Gas left before executing this operation
- `gasCost` Gas cost of this operation
- `stack` Array of all values on the stack
- `depth` Depth of the call stack
- `returnStack` Array of values, Stack of the called function
- `returnData` Data returned by function call
- `refund` Amount of **global** gas refunded
- `memSize` Size of memory array

Optional Fields: 
- `opName` Name of the operation
- `error` Description of an error 
- `memory` Array of all allocated values
- `storage` Array of all stored values

> 

Reviewer 2: I think it would be worthwhile to say that clients SHOULD output the fields in the order they are listed in this EIP.

*Example:* 
```
{"pc":0,"op":96,"gas":"0x2540be400","gasCost":"0x3","memory":"0x","memSize":0,"stack":[],"depth":1,"error":null,"opName":"PUSH1"}
```

The `stack`, `memory` and `memSize` are the values _before_ execution of the op. 
All array attributes (`stack`, `returnStack`, `memory`) MUST be initialized to empty arrays ("stack":[],) NOT to null.

> If `--nomemory` is passed to the CUT, the `memory` field is omitted (but the `memSize` field has to be printed).

Reviewer 2: This implies that specifically `--nomemory` is part of the spec.
```suggestion
If the CUT will not be outputting values for `memory` then the `memory` field is omitted.  This can be either because the CUT does not support this or because it has been configured not to, such as via a `--nomemory` CLI option.
The `memSize` field MUST be present regardless of `memory` support.
```

Clients SHOULD implement a way to disable recording the storage as the stateroot includes all storage updates anyway.

The CUT MUST NOT output a line for the `STOP` operation if an error occured:
*Example:* 
```
{"pc":2,"op":0,"gas":"0x2540be3fd","gasCost":"0x0","memory":"0x","memSize":0,"stack":["0x40"],"depth":1,"error":null,"opName":"STOP"}
```

### Summary and error handling

> At the end of execution, the CUT MUST print some summerical info, this info SHOULD have the following fields. The summary SHOULD be printed to STDOUT.

Reviewer 2: Should this summary field be a single json object like each json trace line? I think it should be specified as such.

> Required Fields: 

Reviewer 2: Please specify the javascript types of the values.

- `stateRoot` Root of the state trie after executing the transaction
- `output` Return values of the function
- `gasUsed` All gas used by the transaction
- `successful` Bool whether transaction was executed successfully

OptionalFields:

> - `time` Time needed to execute the transaction

Reviewer 2: What time unit?  Millis, Micros, or Nanos?

This proposal is heavily influenced by the requirements for [goevmlab](https://github.com/holiman/goevmlab) and our effort to create differential fuzzers for the EVM.

## Backwards Compatibility
<!--All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->
This EIP is fully backward compatible with ethereum as it only introduces a better tracing infrastructure that is optional for clients to implement.

### Clients
This EIP is fully backward compatible with go-ethereum. OpenEthereum, Besu and Nethermind clients would have to change their JSON output of `openethereum-evm` `evmtool` and `nethtest` slightly do adhere to the new and stricter specs. New clients would need to implement this change if they want to be part of the differential fuzzing group.

## Test Cases
<!--Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. Other EIPs can choose to include links to test cases if applicable.-->

> <details><summary>Test cases</summary>

Reviewer 2: There should be tests exercising `memSize`, `refundCounter`, `returnStack`, and `returnData` since they were called out in the prologue.   Referencing a reference test and including the relevant line would satisfy me.

     }
    }
   ]
  }
 }
}
```

```bash
$evm --nomemory --json statetest test.json
{"pc":0,"op":96,"gas":"0x79be38","gasCost":"0x3","memory":"0x","memSize":0,"stack":[],"returnStack":[],"returnData":"0x","depth":1,"refund":0,"opName":"PUSH1","error":""}
{"pc":2,"op":0,"gas":"0x79be35","gasCost":"0x0","memory":"0x","memSize":0,"stack":["0x40"],"returnStack":[],"returnData":"0x","depth":1,"refund":0,"opName":"STOP","error":""}
{"output":"","gasUsed":"0x3","time":166351}
{"stateRoot": "3f7878bb9cd21378f7eb0a2d26e24f11abca4709482cc92e31c1a8a9e4aeeed5"}
{"pc":0,"op":96,"gas":"0x79be38","gasCost":"0x3","memory":"0x","memSize":0,"stack":[],"returnStack":[],"returnData":"0x","depth":1,"refund":0,"opName":"PUSH1","error":""}
{"pc":2,"op":0,"gas":"0x79be35","gasCost":"0x0","memory":"0x","memSize":0,"stack":["0x40"],"returnStack":[],"returnData":"0x","depth":1,"refund":0,"opName":"STOP","error":""}
{"output":"","gasUsed":"0x3","time":50655}
{"stateRoot": "3f7878bb9cd21378f7eb0a2d26e24f11abca4709482cc92e31c1a8a9e4aeeed5"}
[
  {
    "name": "TraceTest",
    "pass": true,
    "fork": "Istanbul"
  },
  {
    "name": "TraceTest",
    "pass": true,
    "fork": "Berlin"
  }
]
```

</details>

> ## Implementation

Reviewer: Note, this section recently became optional. I recommend removing it as we prefer to not have external links in EIPs.

Author: I'm not sure about this. I think it would be good for people to have a look at implementation examples if they want to implement it in their client

Reviewer 3: Implementation examples are good, just not appropriate for a technical specification document.  EIPs should not be seen as a tutorial or educational material, such content is much better suited for blogs, programming websites, etc.

> <!--The implementations must be completed before any EIP is given status "Final", but it need not be completed before the EIP is accepted. While there is merit to the approach of reaching consensus on the specification and rationale before writing code, the principle of "rough consensus and running code" is still useful when it comes to resolving many discussions of API details.-->

Reviewer: ```suggestion
```

Implementation in [go-ethereum](https://github.com/ethereum/go-ethereum/tree/master/cmd/evm)
Implementation in [OpenEthereum](https://github.com/openethereum/openethereum/tree/master/evmbin)
Implementation in [Besu](https://github.com/hyperledger/besu/tree/master/ethereum/evmtool)
Implementation in [Nethermind](https://github.com/NethermindEth/nethermind/tree/master/src/Nethermind/Nethermind.State.Test.Runner)

> ## Security Considerations

Reviewer: I'm not sure what is written here is actually a security consideration. It seems like more motivation.

Author: Moved, except for the last sentence

> <!--All EIPs must contain a section that discusses the security implications/considerations relevant to the proposed change. Include information that might be important for security discussions, surfaces risks and can be used throughout the life cycle of the proposal. E.g. include security-relevant design decisions, concerns, important discussions, implementation-specific guidance and pitfalls, an outline of threats and risks and how they are being addressed. EIP submissions missing the "Security Considerations" section will be rejected. An EIP cannot proceed to status "Final" without a Security Considerations discussion deemed sufficient by the reviewers.-->

Reviewer: ```suggestion
```

At the end of execution, the CUT MUST print some summerical info, this info SHOULD have the following fields. The summary SHOULD be printed to STDOUT.
The summary should be a single jsonl object.

Required Fields: 
| Name  | Type  | Explanation  |
|---|---|---|
| `stateRoot` | Hex-String  | Root of the state trie after executing the transaction  |
| `output` |   | Return values of the function  |
| `gasUsed` | Hex-Number  | All gas used by the transaction  |
| `pass` | Boolean  | Bool whether transaction was executed successfully |

OptionalFields:
| Name  | Type  | Explanation  |
|---|---|---|
| `time` |  Number | Time in nanoseconds needed to execute the transaction  |
| `fork` | String  | Name of the fork rules used for execution  |
```
{"stateRoot":"0xd4c577737f5d20207d338c360c42d3af78de54812720e3339f7b27293ef195b7","output":"","gasUsed":"0x3","successful":"true","time":141485}
```

The state test format allows for multiple `postStates` for different forks.
If different `postStates` are present the CUT MUST execute the test according to the rules of the fork sequentially.
It MUST print a trace for each execution as well as an individual summary for each test.

## Rationale

> This EIP is created to enable easier testing of EVM implementations to create a safer ecosystem. It is largely based on the previous non-official documentation for EVM tracing. It tries to cover as many corner cases as possible to enable true client compatibility.

Reviewer 3: This all sounds like motivation, not Rationale.  Rationale is a place to discuss why specific decisions were made.  Why was a particular name used instead of another, why was a particular constant used instead of another, why was a particular design chosen instead of another, etc.

> This proposal is heavily influenced by the requirements for [goevmlab](https://github.com/holiman/goevmlab) and our effort to create differential fuzzers for the EVM.

Reviewer 3: ```suggestion
```
Technical documents are not the right place for shoutouts, attributions, etc.  Recommend removing this and including it in a blog or news article instead.

{"stateRoot": "3f7878bb9cd21378f7eb0a2d26e24f11abca4709482cc92e31c1a8a9e4aeeed5"}
{"pc":0,"op":96,"gas":"0x79be38","gasCost":"0x3","memory":"0x","memSize":0,"stack":[],"returnStack":[],"returnData":"0x","depth":1,"refund":0,"opName":"PUSH1","error":""}
{"pc":2,"op":0,"gas":"0x79be35","gasCost":"0x0","memory":"0x","memSize":0,"stack":["0x40"],"returnStack":[],"returnData":"0x","depth":1,"refund":0,"opName":"STOP","error":""}
{"output":"","gasUsed":"0x3","time":50655}
{"stateRoot": "3f7878bb9cd21378f7eb0a2d26e24f11abca4709482cc92e31c1a8a9e4aeeed5"}
[
  {
    "name": "TraceTest",
    "pass": true,
    "fork": "Istanbul"
  },
  {
    "name": "TraceTest",
    "pass": true,
    "fork": "Berlin"
  }
]
```

</details>

## Implementation
Implementation in [go-ethereum](https://github.com/ethereum/go-ethereum/tree/master/cmd/evm)
Implementation in [OpenEthereum](https://github.com/openethereum/openethereum/tree/master/evmbin)
Implementation in [Besu](https://github.com/hyperledger/besu/tree/master/ethereum/evmtool)
Implementation in [Nethermind](https://github.com/NethermindEth/nethermind/tree/master/src/Nethermind/Nethermind.State.Test.Runner)

## Security Considerations

> Creating better infrastructure might enable attackers to find consensus issues easier.

Reviewer 3: Security Considerations section is for discussing potential security issues that this EIP may introduce, not for discussing existing security problems that this EIP fixes.  😄 
```suggestion
Generation of these traces are expensive, and if a client exposes them to the general public an attacker may be able to execute a denial of service attack against the client.  Clients should consider putting trace endpoints behind a separate flag from other endpoints.
```
